maven-core/src/main/java/org/apache/maven/lifecycle/internal/MojoDescriptorCreator.java

```diff
@@ -176,18 +176,31 @@ public MojoDescriptor getMojoDescriptor( String task, MavenSession session, Mave
         }
         else if ( numTokens == 3 )
         {
-            // We have everything that we need except the version
-            //
-            // org.apache.maven.plugins:maven-remote-resources-plugin:???:process
-            //
-            // groupId
-            // artifactId
-            // ???
-            // goal
-            //
-            plugin = new Plugin();
-            plugin.setGroupId( tok.nextToken() );
-            plugin.setArtifactId( tok.nextToken() );
+            // groupId:artifactId:goal or pluginPrefix:version:goal (since Maven 3.9)
+
+            String firstToken = tok.nextToken();
+            // groupId or pluginPrefix? heuristics: groupId contains . but not pluginPrefix
+            if ( firstToken.contains( "." ) )
```

Comment:
I don't think we can make this assumption for sure since for testing or other reasong the groupId could be: `foo`, `bar-baz`

Reply:
if you want to test a plugin for such simplified run, just don't define such a groupId
at least, it seems to work for real world plugins, that's the first intent

of course, if you find a better heuristics, or even an algorithm, I'm all ears open

Reply:
I gave it some thought, but I can't find a better heuristic either. The only thing more or less foolproof would be to look at the first token, assuming it is a plugin prefix. If that lookup yields nothing, it would be a groupId.

Even that is not 100% safe, though. Imagine you're not connected to the network, so lookup fails, so we decide it's a groupId -- this certainly doesn't make sense. But more importantly, it's probably quite costly compared to a `String.contains` :-).

Reply:
and do you really think that we'll have any single word groupId in the future, given Maven Central explicitely asks for DNS matching?
heuristics is heuristics: in theory there could be edge cases, but in real world, it simply works